observer/holder/nodesHolder.go

```diff
@@ -0,0 +1,436 @@
+package holder
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+
+	logger "github.com/multiversx/mx-chain-logger-go"
+	"github.com/multiversx/mx-chain-proxy-go/data"
+)
+
+var (
+	errEmptyShardIDsList  = errors.New("empty shard IDs list")
+	errWrongConfiguration = errors.New("wrong observers configuration")
+	log                   = logger.GetOrCreate("observer/holder")
+)
+
+type nodesHolder struct {
+	mut                    sync.RWMutex
+	syncedNodes            []*data.NodeData
+	outOfSyncNodes         []*data.NodeData
+	syncedFallbackNodes    []*data.NodeData
+	outOfSyncFallbackNodes []*data.NodeData
+	lastSyncedNodes        map[uint32]*data.NodeData
+	shardIDs               []uint32
+	availability           data.ObserverDataAvailabilityType
+}
+
+// NewNodesHolder will return a new instance of a nodesHolder
+func NewNodesHolder(syncedNodes []*data.NodeData, fallbackNodes []*data.NodeData, shardIDs []uint32, availability data.ObserverDataAvailabilityType) (*nodesHolder, error) {
+	if len(shardIDs) == 0 {
+		return nil, errEmptyShardIDsList
+	}
+
+	return &nodesHolder{
+		syncedNodes:            syncedNodes,
+		outOfSyncNodes:         make([]*data.NodeData, 0),
+		syncedFallbackNodes:    fallbackNodes,
+		outOfSyncFallbackNodes: make([]*data.NodeData, 0),
+		lastSyncedNodes:        make(map[uint32]*data.NodeData),
+		shardIDs:               shardIDs,
+		availability:           availability,
+	}, nil
+}
+
+// UpdateNodes will update the internal maps based on the provided nodes
+func (nh *nodesHolder) UpdateNodes(nodesWithSyncStatus []*data.NodeData) {
```

Comment:
this method does multiple reads and writes on the slices, it should be protected by mutex as well

Reply:
refactored